### ISSUE TITLE
Local state changes are painful (compared to everything else)

### ISSUE THREAD

**Issue by machineghost** [apollographql/apollo-feature-requests#179]
Overall I've been very impressed with Apollo.  The hook mostly feel slick and work just how I want them to.

But then you get to local data updates and all that goes out the window.  Here's the most basic mutation I can think of: a logout mutation (ie. delete a single "user" value):

    const [logout] = useMutation(LOGOUT, {
        update(cache, { data: { logout: { user } } }) {
          const session = cache.readQuery({ query: SESSION });
          cache.writeQuery({
            query: SESSION,
           data: {...session, user},
         });
       }
    });

Yikes!  All that just to tell Apollo "that mutation you just ran deleted a value".  Why so much?  Why not something like:

      const [logout] = useMutation(LOGOUT, SESSION, (session, data: { logout: { user } } }) =>
         ({...session, user}));

In other words, why not allow a `useMutation` signature of

     useMutation(mutationQuery, dataQuery, (dataFromQuery, dataFromMutation) =>
          newDataForQuery;

**Comment by machineghost** [comment]
A basic proof of concept implementation (I can submit a real PR if desired):

    const useMutation2 = (mutation, query, reducer) => {
      const update = (cache, { data }) => {
        const queryData = cache.readQuery({ query });
        cache.writeQuery({ data: reducer(queryData, data), query });
      }
      return useMutation(mutation, { update });
    };
